### ISSUE TITLE
Show on App Open Only!

### ISSUE THREAD

**Issue by ItzNotABug** [ItzNotABug/AppOpenAdManager#5]
It can be useful if there's the control with the developer to show the Ad **only** when the App is opened & **not** when the app is resumed (jump to app from, say, recent apps.)

**Comment by ItzNotABug** [comment]
Added in the `dev` branch,
It is **highly recommended** to use a `SplashScreen` flow, along with a `noHistory` tag in the Manifest to get the expected result.

**Calling `this.finish()` in the SplashActivity, right after calling `startActivity()` will result in not showing the Ad.**

Check the [sample app](https://github.com/ItzNotABug/AppOpenAdManager/tree/master/app).

**Comment by amarilindra** [comment]
Not working. Where am I doing it wrong? 

**Application Class**

```
ArrayList<Class<? extends Activity>> activities = new ArrayList<>();
        activities.add(SplashScreenActivity.class);
        AppOpenAdManager adManager = AppOpenAdManager.get(
                this,
                new Configs(
                        InitialDelay.NONE,
                        application.getResources().getString(R.string.admob_app_open_ad),
                        new AdRequest.Builder().build(),
                        false,
                        null,
                        activities,
                        AppOpenAd.APP_OPEN_AD_ORIENTATION_PORTRAIT
                )
        );
        adManager.loadAppOpenAd();
```

**Splash Screen:**

```
   new Handler().postDelayed(() -> {
            Intent i = new Intent(mActivity, HomeActivity.class);
            startActivity(i);
        }, 1000);

```

**Manifest**

```
  <activity
            android:name="com........SplashScreenActivity"
            android:noHistory="true"
            android:exported="true">
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />
                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>
```

**Comment by ItzNotABug** [comment]
I tried your example & it seems to work.
However that `postDelayed` can cause a different problem (not related to this one though).

Can you tell me how do you close & re-open your app?
Remember that the `SplashScreenActivity` must be opened from a hot-start & the Ad should be loaded.
The Ad would mostly not be available on a cold-start.

**Comment by amarilindra** [comment]
**It was working when:**

1. User opens the app 
2. Splash Screen is set to wait for 10 seconds (only for testing)
3. User minimizes the app
4. User reopens the app after 2 seconds 
5. App Open Ad is shown as expected 


**Useful Usecase for many developers:**
1. User opens the app 
2. Splash Screen is set to wait for 2 seconds 
3. If the App Open Ad is shown within 2 seconds. Show the Ad and wait till the user closes the ad. Now open the MainActivity 
4. If the App Open Ad fails to load within 2 seconds, Open MainActivity and App Open Ad is never shown

**Why this use case:**
The banner Ad will be shown on most of the screens and showing App Open Ad over a Banner Ad is a clear violation. Since this affects a lot of small apps, it would help developers if an App Open ad is shown immediately on the Splash Screen.

**Comment by ItzNotABug** [comment]
Since the AppOpenAd is shown on a hot-start & if you check the sample MainActivity,
You'll see I've used `finish()` in the `onBackPressed()` to kill that Activity which will open the `SplashScreenActivity` on the next hot-start. This was the strategy used to show the Ad only on Splash screens.

Since your use case includes _waiting_ for some time & then navigating to the MainActivity, I'd suggest implementing your logic with the `AppOpenAdListener`.

Why I would not suggest using `postDelayed` in SplashActivity is because (this is what I've usually observed),
1. Assuming the Ad is loaded & shown on Splash
2. SplashActivity completes its `postDelayed` during this period & navigates to MainActivity
3. If you close/pause your MainActivity & exit the app, you'll see the AppOpenAd still not dismissed & is now _**outside**_ the App which is a violation.

**Comment by amarilindra** [comment]
> 1. Assuming the Ad is loaded & shown on Splash
> 2. SplashActivity completes its postDelayed during this period & navigates to MainActivity
> 3. If you close/pause your MainActivity & exit the app, you'll see the AppOpenAd still not dismissed & is now outside the App which is a violation.

I think this is not a problem. We can set a flag when the App Open Ad is shown and we'll check it in `postDelayed` callback before opening the new intent. If the flag is true, don't open the next activity. 

Using `AppOpenAdListener` and `onAdDismissedFullScreenContent`, we can start the new intent. Alternatively, as you suggested completely remove the postDelayed and make use of `onAdFailedToShowFullScreenContent` to open new Intent when App Open Ad is not shown.

**Now the real question. Is it possible to listen to `AppOpenAdListener` events on SplashScreen with this library?**

**Comment by ItzNotABug** [comment]
Yes, you can absolutely listen to the events of `AppOpenAdListener` in SplashScreen.
You can do something like this (example is in Kotlin but its easy to figure out):

```kotlin
class BaseApp : Application() {

    lateinit var appOpenAdInstance: AppOpenAdManager

    override fun onCreate() {
        super.onCreate()
        // Load & show the Ad on a Cold start.
        val configs = Configs(showAdOnFirstColdStart = true)
        appOpenAdInstance = AppOpenAdManager.get(this, configs)
    }
}
```

then in SplashScreen, you can access the `appOpenAdInstance` like below:
```kotlin
class SplashActivity : Activity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        attachAdListener()
        // other things...
    }

    private fun attachAdListener() {
        // We know that the application is a `BaseApp` instance.
        val appOpenAdManager = (application as BaseApp).appOpenAdManager
        appOpenAdManager.setAppOpenAdListener(object : AppOpenAdListener() {
            override fun onAdLoaded() { println("onAdLoaded") }
            override fun onAdWillShow() { println("onAdWillShow") }
            override fun onAdShown() { println("onAdShown") }

            override fun onAdDismissed() {
                // Ad is dismissed, move to MainActivity.
                startActivity(Intent(this@SplashActivity, MainActivity::class.java))
            }

            override fun onAdShowFailed(error: AdError?) {
                // Failed to show the Ad, handle the AdError & move to MainActivity.
                startActivity(Intent(this@SplashActivity, MainActivity::class.java))
            }

            override fun onAdFailedToLoad(loadAdError: LoadAdError) {
                // Failed to load the Ad, handle the LoadAdError & move to MainActivity.
                startActivity(Intent(this@SplashActivity, MainActivity::class.java))
            }
        })
    }
}
```
